### ISSUE TITLE
Bug: Dependencies Array shows missing dependency error for generic type parameters

### ISSUE THREAD

**Issue by mrousavy** [facebook/react#19808]
React version: 16.13.1

## Steps To Reproduce

1. Create function with generic type parameter
2. Use that generic type parameter in `useCallback`, `useEffect` or `useMemo`

Example:

```ts
export function useDocument<TDocument>(path: string, defaultValue: TDocument | undefined = undefined): TDocument | undefined {
	const [value, setValue] = useState<TDocument | undefined>(defaultValue);

	const onNext = useCallback((snapshot: FirebaseFirestoreTypes.DocumentSnapshot<FirebaseFirestoreTypes.DocumentData>) => {
		if (snapshot.exists) setValue(build<TDocument>(snapshot.data(), snapshot.id));
		else setValue(undefined);
	}, []);


	useEffect(() => firestore().doc(path).onSnapshot(onNext), [onNext, path]);

	return value;
}
```

![Screenshot 2020-09-10 at 19 57 06](https://user-images.githubusercontent.com/15199031/92776710-ce9d8200-f39f-11ea-84bb-3d03fb76f8f1.png)

## The current behavior

Shows error `React Hook useCallback has a missing dependency: 'TDocument'. Either include it or remove the dependency array.eslintreact-hooks/exhaustive-deps)`

## The expected behavior

Should not show any errors, because it isn't possible to add types to the dependencies array (types don't exist at runtime!)

**Comment by dsgkirkby** [comment]
I'm also seeing this issue. It only happens after updating typescript-eslint from v3.8 to v4.1.

**Comment by mrousavy** [comment]
@dsgkirkby yep, might be a typescript 4.0 problem

**Comment by eps1lon** [comment]
Duplicate of #19742. It should be fixed by #19751. We'll just need to wait for a release.

**Comment by gaearon** [comment]
`eslint-plugin-react-hooks@4.1.1` should have the fix.
